Add JDBC

NikishinEvgeniy opened the pull request:
1. Начал юзать ломбок ( работа над ошибками)
2. Начал юзать паттерн Builder ( для создания сложных объектов)

NikishinEvgeniy committed:
Add JDBC

1. Начал юзать ломбок ( работа над ошибками)
2. Начал юзать паттерн Builder ( для создания сложных объектов)

YlabCourse commented on homework/README.md:
должна быть сборка приложения и создания docker образа вашего приложения.

YlabCourse commented on homework/docker-compose.yaml:
```diff
@@ -0,0 +1,12 @@
+
+version: '3.3'
+services:
```
добавить ваш сервис и dockerfile создания образа

YlabCourse commented on homework/src/main/java/carshop_service/application/ConfigLoader.java:
```diff
@@ -0,0 +1,55 @@
+    private Properties properties = new Properties();
+
+    public ConfigLoader(String configFileName) {
+        try (InputStream input = getClass().getClassLoader().getResourceAsStream(configFileName)) {
+            if (input == null) {
+                throw new RuntimeException("Не удалось найти конфигурационный файл: " + configFileName);
+            }
+            properties.load(input);
+            this.url = getProperty("car_service.url");
+            this.password = getProperty("car_service.password");
+            this.username = getProperty("car_service.username");
+        } catch (IOException e) {
+            throw new RuntimeException("Ошибка загрузки конфигурационного файла: " + configFileName, e);
+        }
```
лучше разделить на разные классы загрузку пропертей и конфигурацию базы данных

YlabCourse commented on homework/src/main/java/carshop_service/application/LiquidBaseInitialize.java:
```diff
@@ -0,0 +1,33 @@
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+public class LiquidBaseInitialize {
```
классы называют существительными, а методы глаголами

YlabCourse commented on homework/src/main/java/carshop_service/application/LiquidBaseInitialize.java:
```diff
@@ -0,0 +1,33 @@
+public class LiquidBaseInitialize {
+    private static ConfigLoader configLoader;
+
+    public static void main(String[] args) {
+        configLoader = new ConfigLoader("application.properties");
+        try {
+            Connection connection = DriverManager.getConnection(configLoader.getProperty("liquibase.url"),
+                    configLoader.getProperty("liquibase.username"),configLoader.getProperty("liquibase.password"));
+            Database database =
+                    DatabaseFactory
+                            .getInstance()
+                            .findCorrectDatabaseImplementation(new JdbcConnection(connection));
+            Liquibase liquibase
+                    = new Liquibase(configLoader.getProperty("liquibase.changelog"), new ClassLoaderResourceAccessor(),database);
+            liquibase.update();
+        }
+        catch (SQLException | LiquibaseException e) {
+            throw new RuntimeException(e);
+        }
+    }
```
почему миграция в отдельном приложении

YlabCourse commented on homework/src/main/java/carshop_service/dao/ClientDaoImpl.java:
```diff
@@ -1,48 +1,184 @@
+        PreparedStatement statement = null;
+        try {
+            configLoader.setSchema(connection.createStatement());
+            String addClientQuery = "INSERT INTO client (name,surname,login,password,role,state,count_of_buy) " +
```
sql запросы лучше вынести в константы класса

YlabCourse commented on homework/src/main/java/carshop_service/dao/ClientDaoImpl.java:
```diff
@@ -1,48 +1,184 @@
+        }
+        finally {
+            try {
+                connection.close();
```
Statement и ResultSet тоже надо закрывать

YlabCourse commented on homework/src/main/java/carshop_service/entity/OrderBuilder.java:
Билдеры можно делать внутренними классами сущностей, или использовать аннотацию от lombok

YlabCourse commented on homework/src/main/java/carshop_service/exception/NoSuchCarException.java:
все исключения об отсутствии сущностей можно объединить в одно и просто передавать в него описание.

YlabCourse commented on homework/src/main/resources/database/001-create-table.xml:
```diff
@@ -0,0 +1,60 @@
+    <changeSet id="001-create-table" author="evgen">
+
+        <createTable tableName="car" schemaName="private_schema">
+            <column name="id" type="SERIAL">
```
на будущее используйте remark

YlabCourse commented on homework/src/main/resources/log.txt:
логи не должны попадать в git репозиторий

YlabCourse commented on homework/src/test/java/carshop_service/dao/CarDaoTest.java:
```diff
@@ -1,48 +1,91 @@
+            = new PostgreSQLContainer<>("postgres:latest")
+            .withDatabaseName("car_serivce")
+            .withUsername("evgen")
+            .withPassword("admin");
```
контейнер можно вынести в отдельный класс конфигурации в статическое поле, чтобы использовать один контейнер на все тесты с базой, это ускорит их выполнение

NikishinEvgeniy committed:
work on bugs

1. Исправил код с JDBC ( Connection,Statment,ResultSet) закрываются в коде
2. Сделал 2 класса. Считывание проперти из проперти файла. Конфигурация базы данных. Раньше все было в 1 классе
3. Переименовал класс с LiquidBaseInitialize -> LiquidBaseInitializer
4. Вынес SQL запросы в константы класса
5. Сделал Builder внутренними классами + использовал аннотацию от ломбок builder
6. Вынес PostgresContainer в отдельный класс ( для тестов )
7. Добавил remarks в liquidBase xmlфайлы

Осталось:
1. Собрать приложение полностью в Docker (Postgres,LiquidBase,Application)
2. Создать  приватную схему, для служебных таблиц
3. Поправить файл с логами, чтобы не падал в git
